src/terminal/input/terminalInput.cpp

```diff
@@ -515,169 +162,442 @@ TerminalInput::OutputType TerminalInput::HandleKey(const INPUT_RECORD& event)
+    const auto flags = 4u; // Don't modify the state in the ToUnicodeEx call.
+    const auto bufferSize = 16;
+    auto buffer = std::array<wchar_t, bufferSize>{};
+
+    // However, we first need to query the key with the original state, to check
+    // whether it's a dead key. If that is the case, ToUnicodeEx should return a
+    // negative number, although in practice it's more likely to return a string
+    // of length two, with two identical characters. This is because the system
+    // sees this as a second press of the dead key, which would typically result
+    // in the combining character representation being transmit twice.
+    auto length = ToUnicodeEx(virtualKeyCode, 0, keyState.data(), buffer.data(), bufferSize, flags, hkl);
+    if (length < 0 || (length == 2 && buffer.at(0) == buffer.at(1)))
     {
-        if (auto out = _searchWithModifier(keyEvent))
+        return MakeOutput({});
+    }
+
+    // Once we know it's not a dead key, we run the query again, but with the
+    // Ctrl and Alt modifiers disabled to obtain the base character mapping.
+    keyState.at(VK_CONTROL) = keyState.at(VK_LCONTROL) = keyState.at(VK_RCONTROL) = 0;
+    keyState.at(VK_MENU) = keyState.at(VK_LMENU) = keyState.at(VK_RMENU) = 0;
+    length = ToUnicodeEx(virtualKeyCode, 0, keyState.data(), buffer.data(), bufferSize, flags, hkl);
+    if (length > 0)
+    {
+        auto charSequence = StringType{ buffer.data(), gsl::narrow_cast<size_t>(length) };
+        // Once we've got the base character, we can apply the Ctrl modifier.
+        if (ctrlIsReallyPressed && charSequence.length() == 1)
         {
-            return out;
+            charSequence.at(0) = _makeCtrlChar(charSequence.at(0));
+            // If we haven't found a Ctrl mapping for the key, and it's one of
+            // the alphanumeric keys, we try again using the virtual key code.
+            // On keyboard layouts where the alphanumeric keys are not mapped to
+            // their typical ASCII values, this provides a simple fallback.
+            if (charSequence.at(0) >= L' ' && virtualKeyCode >= '2' && virtualKeyCode <= 'Z')
+            {
+                charSequence.at(0) = _makeCtrlChar(virtualKeyCode);
+            }
+        }
+        // If Alt is pressed, that also needs to be applied to the sequence.
+        if (altIsPressed)
+        {
+            charSequence = _makeEscapedOutput(charSequence);
         }
+        return charSequence;
     }
 
-    // This section is similar to the Alt modifier section above,
-    // but handles cases without Ctrl modifiers.
-    if (WI_IsAnyFlagSet(keyEvent.dwControlKeyState, ALT_PRESSED) && !WI_IsAnyFlagSet(keyEvent.dwControlKeyState, CTRL_PRESSED) && keyEvent.uChar.UnicodeChar != 0)
+    // If we've got nothing usable, we'll just return an empty string. The event
+    // has technically still been handled, even if it's an unmapped key.
+    return MakeOutput({});

```

Comment:
It's just a style nit, but it may be neat to flip the above `length > 0` check, because then every `return MakeOutput({})` is an early return in this function.